compiler/src/dotty/tools/dotc/transform/sjs/ExplicitJSClasses.scala

```diff
@@ -12,54 +17,728 @@ import core.Types._
+ *  instead declared as a `var` to work around the cyclic dependency:
+ *  {{{
+ *  def outer() = {
+ *    var Local\$jsclass: AnyRef = null
+ *    class Local extends ParentJSClass {
+ *      def newLocal = new Local // self-reference
+ *    }
+ *    Local\$jsclass = createLocalJSClass(...)
+ *  }
+ *  }}}
+ *
+ *  --------------------------------------
+ *
+ *  (D) Insertion of `withContextualJSClassValue` calls
+ *
+ *  For any nested JS class `C`, this phase performs the following
+ *  transformations:
+ *
+ *  - `new C[...Ts](...args)` desugars into
+ *    `withContextualJSClassValue(js.constructorOf[C], new C[...Ts](...args))`,
+ *    so that the back-end receives a reified reference to the JS class value.
+ *  - In the same spirit, for `D extends C`, `D.super[C].m[...Ts](...args)`
+ *    desugars into
+ *    `withContextualJSClassValue(js.constructorOf[C], D.super[C].m[...Ts](...args))`.
+ *
+ *  For any nested JS *object*, their (only) instantiation point of the form
+ *  `new O$()` is rewritten as
+ *  `withContextualJSClassValue(js.constructorOf[ParentClassOfO], new O$())`,
+ *  so that the back-end receives a reified reference to the parent class of
+ *  `O`.
+ *
+ *  TODO A similar treatment is applied on anonymous JS classes, which
+ *  basically define something very similar to an `object`, although without
+ *  its own JS class.
+ *
+ *  --------------------------------------
+ *
+ *  (E) Desugar `x.isInstanceOf[C]` for nested JS classes
  *
- *  - Rewrite `js.constructorOf[T]` into `scala.scalajs.runtime.constructorOf(classOf[T])`,
- *    where the `classOf[T]` is represented as a `Literal`.
+ *  They are desugared into `js.special.instanceof(x, js.constructorOf[C])`.
+ *
+ *  --------------------------------------
+ *
+ *  (F) Desugar `js.constructorOf[C]`
+ *
+ *  Finally, this phase rewrites all calls to `js.constructorOf[C]`, including
+ *  the ones generated by the previous steps. The transformation depends on the
+ *  nature of `C`:
+ *
+ *  - If `C` is a statically accessible class, desugar to
+ *    `runtime.constructorOf(classOf[C])` so that the reified symbol survives
+ *    erasure and reaches the back-end.
+ *  - If `C` is an inner JS class, it must be of the form `path.D` for some
+ *    pair (`path`, `D`), and we desugar it to `path.D\$jsclass`, using the
+ *    field created by step (A) (it is an error if `C` is of the form
+ *    `Enclosing#D`).
+ *  - If `C` is a local JS class, desugar to `C\$jsclass`, using the local val
+ *    created by step (C).
  */
 class ExplicitJSClasses extends MiniPhase with InfoTransformer { thisPhase =>
   import ExplicitJSClasses._
   import ast.tpd._
 
   override def phaseName: String = ExplicitJSClasses.name
 
+  private var MyState: Store.Location[MyState] = _
+  private def myState(using Context) = ctx.store(MyState)
+
+  override def initContext(ctx: FreshContext): Unit =
+    MyState = ctx.addLocation[MyState]()
```

Comment:
@gzm0 This is standard in practice in dotty ;) I copied the pattern from `LambdaLift`.